crates/transport/src/util/stream_with_payload.rs

```diff
@@ -0,0 +1,70 @@
+use std::{
+    io::{IoSlice, IoSliceMut},
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures::{AsyncRead, AsyncWrite};
+
+#[pin_project::pin_project]
+pub struct StreamWithPayload<S, P> {
+    #[pin]
+    stream: S,
+    payload: P,
```

Comment:
So the `payload` here is actually just a semaphore permit? And the whole struct exists to drop the permit after the stream is done?

Reply:
Yes, exactly! I would gladly use something existing from `futures::stream::StreamExt`, but couldn't find anything suitable

Reply:
It would be great if there was an existing create for this. Seems like a quite common use case. This code does the job perfectly, but it's a bit verbose. I was thinking maybe instead of implementing the traits for `StreamWithPayload` itself, you could implement `Deref` and `DerefMut` to allow calling methods on the `stream` field? (BTW, this could be derived with `derive_more` crate.)

Reply:
Implementing `Deref` was a good idea, but it turned out it's not enough for moves:
```rust
cannot move out of dereference of `impl DerefMut + Deref<Target = impl futures::AsyncRead + futures::AsyncWrite>`
...
stream.take(self.config.max_response_size + 1).read_to_end(&mut buf).await?;
```

There is [this crate](https://docs.rs/stream-guard/latest/stream_guard/) for `futures::Stream`, but I think converting between streams and AsyncRead/Write would be even more awkward. Couldn't find any suitable functions for the AsyncRead/Write yet. There is no even `map` so that I could stuff this guard into the closure :(

Reply:
Actually, you can quite easily convert `AsyncRead` into `Stream` with [ReaderStream](https://docs.rs/tokio-util/latest/tokio_util/io/struct.ReaderStream.html). And for `AsyncWrite` there's [into_sink](https://docs.rs/futures/latest/futures/io/trait.AsyncWriteExt.html#method.into_sink) method.

Reply:
Yes, but here we have a single type which is both `AsyncRead` and `AsyncWrite`, so it's not convertible to either of those

Reply:
I see, in that case it's probably best to leave it as is.